cilium/CFP-36372-neigh-entry-refresh-non-cluster.md

```diff
@@ -0,0 +1,136 @@
+non-cluster IPs in a stale state indefinitely. In particular for node port
+traffic, Cilium calls `bpf_fib_lookup()` which returns a valid neighbor entry
+even if it is stale (returning `BPF_FIB_LKUP_RET_SUCCESS`), populates the source
+and destination MAC addresses from the fib lookup result, then
+[calls `bpf_redirect()`](https://github.com/cilium/cilium/blob/a184399ad122b293f531e56293e5cd67eca4622a/bpf/lib/fib.h#L171)
+sending the packet to interface returned by `bpf_fib_lookup()`. `bpf_redirect()`
+does not trigger `neigh_event_send_probe()` as the kernel stack or
+`bpf_redirect_neigh()` would. Cilium only opts for `bpf_redirect_neigh()` in
+cases where the preceding `bpf_fib_lookup()` returns `BPF_FIB_LKUP_NO_NEIGH`
+which is not the case for stale entries.
+
+The net result is that stale neighbor entries for non-cluster IPs remain stale
+unless it is deleted/GCed, an ARP request is sent from the neighbor in question,
+or the entry is updated externally. This becomes problematic when an IP address
+is reassigned and mapped to a new MAC address, a common scenario in virtualized
+environments with frequent node churn, as traffic will never be able to make its
+way from cluster nodes to the new owner of that IP. Stale entries may never be
+garbage collected depending on the number of entries in the neighbor table and
+the [configured GC thresholds](https://man7.org/linux/man-pages/man7/arp.7.html)
+(`gc_thresh1`, `gc_thresh2`, `gc_thresh3`). In environments where ARP
+suppression is enabled, common with vxlan setups, the switch responds directly
+to ARP requests and can suppress gratuitous ARPs broadcast by the neighbor that
+would otherwise inform cluster nodes of the IP reassignment. As of today, Cilium
+does not have a mechanism in place to keep neighbor entries fresh in such
+scenarios with BPF host routing.
+
+## Goals
+
+* Ensure that neighbor entries remain up-to-date for non-cluster IPs when using
+  BPF host routing.
+
+## Non-Goals
+* __
+
+## Proposal
+
+### Overview
+This proposal compares three alternative solutions one of which involves a
+kernel change.
+
+### Option 1: `managed` entries for custom custom IPs
+Create `managed` neighbor entries for a static set of custom IP addresses
+provided to Cilium by some flag (e.g. `--managed-l2-neighbors=192.168.1.88,...`).
+If the set of non-cluster IPs is small and known up front this is a simple
+extension to Cilium's existing logic which creates such entries for node IPs.
+Actively monitoring and removing stale entries for these addresses also achieves
+the same goal.
+
+#### Pros
+* Requires no changes to the kernel.
+* Easy to implement
+
+#### Cons
+* Not very flexible and harder to manage. The feedback from our team internally
+  was that in their situation it is hard to know ahead of time which IPs will be
+  used.
+
+### Option 2: Always use `bpf_redirect_neigh()`
+Always use `bpf_redirect_neigh()` even if `bpf_fib_lookup()` returns
```

Comment:
Is there a particular set of conditions that make using the stale entry provided by fib lookup more likely?  Maybe this could be an opt-int option rather than making it the default (i.e. to balance performance vs correctness)?

Reply:
> Is there a particular set of conditions that make using the stale entry provided by fib lookup more likely?

Assuming this behavior is opt-in, this would apply to any environments where a user has some host(s) talking to cluster nodes in the same L2 domain and where the IP address of those external hosts may migrate to new hosts.

I believe our specific scenario involved a virtual load balancer appliance talking to a node port service where both the appliance and cluster nodes are VMs. In some scenarios (environment upgrades) these VMs are replaced and IPs reassigned to their replacements breaking connectivity.

> Maybe this could be an opt-int option rather than making it the default (i.e. to balance performance vs correctness)?

This is certainly an option, although I worry about adding yet another flag or option for users to dig through to understand if it applies to their scenario. Hopefully we can just have performance *and* correctness. Perhaps some benchmarking is warranted here to determine what the effects would be of always using `bpf_redirect_neigh()` if we don't already have some numbers available. My hope would be that the performance difference is negligible.